apps/cmp.c

```diff
@@ -2990,19 +3012,27 @@ int cmp_main(int argc, char **argv)
         if (!ret)
             goto err;
         ret = 0;
+        if (opt_srvcertout != NULL) {
+            X509 *srvcert = OSSL_CMP_CTX_get0_validatedSrvCert(cmp_ctx);
+
+            if (srvcert == NULL) {
+                if (unlink(opt_srvcertout) != 0 && errno != ENOENT) {
+                    CMP_err1("Failed to delete %s, which should be done to indicate there is no validated server cert",
+                             opt_srvcertout);
```

Comment:
I am not sure this is really needed. We do not delete files in other applications when we indicate an output via some option but there is none.

Reply:
Here the point is that the information whether the server cert has been successfully validated my be critical.
If the user specified `-srvcertout` and the server was not successfully validated in the current run of the client but the given file is still present from an earlier run, this would be misleading.

Actually, the same holds also for the `certout` option.
This may be seen as a little less critical, but I'm going to add the file deletion there, too. How if I do in a separate commit within this PR? This will amount just to a few lines of code.
I checked also about `-chainout` and `-cacertsout`: for these, if requested a file will be produced/updated also in case no such cert is available. In this case the fill will be empty, so no confusion with older file contents is possible.

Could well be that a similar mechanism would be needed also for other options of other applications.

Reply:
I meanwhile realized that 
* the `save_cert()` helper function should better be generalized to include the file deletion if needed, such that this feature takes over also to further certs like the newly enrolled cert, which simplifies its uses
* the file with the newly enrolled cert and the validated server cert (if requested) should also be removed on any client error
* the diagnostics of the `save_free_certs()` helper function deserved some improvements

Done now.